lib/foreman/renderer/errors.rb

```diff
@@ -0,0 +1,35 @@
+module Foreman
+  module Renderer
+    module Errors
+      class RenderingError < Foreman::Exception
+        def initialize(**params)
+          super(N_(self.class::MESSAGE), params)
+        end
+      end
+
+      class SyntaxError < RenderingError
+        MESSAGE = 'Syntax error occurred while parsing the template %{name}, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: %{message}'.freeze
+      end
+
+      class WrongSubnetError < RenderingError
+        MESSAGE = '%{object_name} is a %{object_class}, expected a subnet'.freeze
+      end
+
+      class HostUnknown < RenderingError
+        MESSAGE = 'This templates requires a host to render but none was specified'.freeze
+      end
+
+      class HostParamUndefined < RenderingError
+        MESSAGE = 'Parameter %{name} is not set for host %{host}'.freeze
+      end
+
+      class HostENCParamUndefined < RenderingError
+        MESSAGE = 'Parameter %{name} is not set in host %{host} ENC output, resolving failed on step %{step}'.freeze
+      end
+
+      class FilteredGlobalSettingAccessed < RenderingError
+        MESSAGE = 'Global setting "%s" is not accessible in safe-mode'.freeze
+      end
+    end
+  end
+end
```

Comment:
I am not sold in creating those error classes, it's a standard practice in Java-like languages but in Ruby comparing actual message is not big deal for catching those in test expectations (source string is stored in `Foreman::Exception` unchanged).

All classed defined in this file are only raised in one place and never being rescued in a block (except in tests). Therefore I suggest to only define `RenderingError` if you really think this is needed, but I'd stick with just generic `Foreman::Exception`.

This will also keep the message with interpolation keys together in one place with interpolation hash.

Reply:
I think this is good improvement for user experience. We can have clear errors for known rendering problems. The more macros we'll add, the more we need this.